### ISSUE TITLE
Encoder options for record extension

### ISSUE THREAD

**Issue by atomictag** [kriszyp/cbor-x#48]
Thank you so much for this awesome library (and for the integration with the lmdb package).

I have a question regarding Encoder setup with records / `addExtension` (which works great) and serialization of `Map` objects
Basically I want to be serialize / deserialize custom records (= classes) preserving `Map` instances and not converting them into plain objects. For this to work I have to explicitly pass `{ mapsAsObjects: true }` to the Encoder used by extensions (which has the record extension enabled automatically as the docs say) otherwise `Map` objects are serialized as plain objects. The docs say _"This is disabled by default if useRecords is enabled (Maps are preserved since they are distinct from records)"_ which seem to imply that `useRecords` preserves `Maps` by default, but that does not seem the case. Also, it seems counter-intuitive to set `{ mapsAsObjects: true }` to produce the effect of _not_ converting `Map` to objects (maybe it's a naming issue?). That made me think there might be some option-handling issue. Lastly, the docs state that _"structured cloning also enables preserving certain typed objects like Error, Set, RegExp and TypedArray instances, using registered CBOR tag extensions"_. However, this does not seem to be needed to encode `Set` etc. with a new Encoder (I was trying to figure out if enabling this option was going to preserve `Map` but it doesn't) even though the option is not set at all, AFAICT. 

Many thanks!

**Comment by kriszyp** [comment]
@atomictag I think what might be the problem is how the encode is recursively called from the encode extension. Are you calling the encode function that is provided as the second argument to your extension's encode? If you call the default exported `encode`, this won't have your encoder settings (which is why you probably have to use `mapAsObjects: true`. And if you call the your own encoder instance's `encode`, it will start a new serialization. I added an example/test here https://github.com/kriszyp/cbor-x/commit/b5f37134765453c2e4d099c921c1b8761c9aa6f9 that should demonstrate encoding with preservation of `Map` instances and recursively encoding with the `encode` callback provided to the extension. Note that I am not turning on `mapsAsObjects`, it is left off/false to preserve Map instances with extension. Does that make sense?

**Comment by atomictag** [comment]
Thanks for looking into this.

I'm actually using Typescript - and the `Extension` type does not export the `encode` function with the 2nd argument:

```
interface Extension {
	Class: Function
	tag: number
	encode(value: any): Buffer | Uint8Array
	decode(messagePack: Buffer | Uint8Array): any
}
```
so I resort to the following:

```
let encoder = new Encoder({ mapsAsObjects: true }) // <== this is needed (as per issue)
addExtension({
	Class: Extended,
	tag: 301,
	decode: function(data) {
		let e = new Extended()
		e.value = data[0]
		e.map = data[1]
		return e
	},
	encode: function(instance) {
		return encoder.encode([instance.value, instance.map]) // <== use "custom" encoder to encode
	}
})
var serialized = encode(data) // <== use "default" encode / decode here. encoder.encode(...) won't work!
var deserialized = decode(serialized)
```
My understanding was that the default `encode(data)` and `decode(serialized)` were going to figure out registered extensions for the tags they find and invoke the corresponding decode/encode functions, with the latter using a specialized encoder in its implementation.

The documentation actually [shows](https://github.com/kriszyp/cbor-x#custom-extensions) the 2-arguments `encode` should be used as you described, but I guess I trusted the Typescript interface although type definitions seemed to be a little out of sync with the code.

**Comment by kriszyp** [comment]
Heh, actually just merged a PR to fix that! https://github.com/kriszyp/cbor-x/pull/47
I guess I should publish it!

**Comment by kriszyp** [comment]
> My understanding was that the default encode(data) and decode(serialized) were going to figure out registered extensions for the tags they find and invoke the corresponding decode/encode functions, with the latter using a specialized encoder in its implementation.

The problem with doing this is that independent instances can't efficiently coordinating continuous writing to the same buffer. Does your code work as expected with the second arg encode (and using encoder.encode for the root serialization) like in my example/test?

**Comment by atomictag** [comment]
Awesome. I've replaced the type definitions with those in #47 and adjusted the code and it now works perfectly.
I'm still not 100% sure what's happening behind the scenes (there's quite some complexity at plays here), but this pattern definitely makes sense and it works! 

> The problem with doing this is that independent instances can't efficiently coordinating continuous writing to the same buffer.

I've actually been hitting some odd behavior indeed in some cases with a lot of serializations / deserializations of complex class instances where at times `encode(...)` would return a buffer with length zero. I spent quite some time debugging this but never got to the bottom of it as the problem seemed non-deterministic (e.g, it was happening if I ran a bunch of test cases in series, but it would not happen if I ran each individual test on its own or in smaller test batches). It looked like the buffer was reset just before returning to the caller. My "fix" was to detect those cases and retry the serialization - which would always succeed at that point. After I applied the suggested change I don't see this problem any longer. 

> Heh, actually just merged a PR to fix that! https://github.com/kriszyp/cbor-x/pull/47

While at it I noticed the options in the type definition don't match completely the docs:

=> in type definition but not in docs 
- `encodeUndefinedAsNil` 
- `maxSharedStructures`
- `maxOwnStructures`
- `useSelfDescribedHeader`

=> in docs but not in type definition
- `sequential`
- `pack`
- `useTag259ForMaps`
- `int64AsNumber`
